src/client/clientenvironment.cpp

```diff
@@ -194,32 +194,43 @@ void ClientEnvironment::step(float dtime)
 		lplayer->applyControl(dtime_part, this);
 
 		// Apply physics
-		if (!free_move && !is_climbing) {
+		if (!free_move) {
 			// Gravity
 			v3f speed = lplayer->getSpeed();
-			if (!lplayer->in_liquid)
+			if (!is_climbing && !lplayer->in_liquid)
 				speed.Y -= lplayer->movement_gravity *
 					lplayer->physics_override_gravity * dtime_part * 2.0f;
 
 			// Liquid floating / sinking
-			if (lplayer->in_liquid && !lplayer->swimming_vertical &&
+			if (!is_climbing && lplayer->in_liquid &&
+					!lplayer->swimming_vertical &&
 					!lplayer->swimming_pitch)
 				speed.Y -= lplayer->movement_liquid_sink * dtime_part * 2.0f;
 
-			// Liquid resistance
-			if (lplayer->in_liquid_stable || lplayer->in_liquid) {
-				// How much the node's viscosity blocks movement, ranges
-				// between 0 and 1. Should match the scale at which viscosity
+			// Movement resistance
+			if (lplayer->move_resistance > 0) {
+				// How much the node's move_resistance blocks movement, ranges
+				// between 0 and 1. Should match the scale at which liquid_viscosity
 				// increase affects other liquid attributes.
-				static const f32 viscosity_factor = 0.3f;
-
-				v3f d_wanted = -speed / lplayer->movement_liquid_fluidity;
+				static const f32 resistance_factor = 0.3f;
+
+				v3f d_wanted;
+				bool in_liquid_stable = lplayer->in_liquid_stable || lplayer->in_liquid;
+				if (in_liquid_stable) {
+					d_wanted = -speed / lplayer->movement_liquid_fluidity;
+				} else {
+					d_wanted = -speed / BS;
+				}
 				f32 dl = d_wanted.getLength();
-				if (dl > lplayer->movement_liquid_fluidity_smooth)
-					dl = lplayer->movement_liquid_fluidity_smooth;
-
-				dl *= (lplayer->liquid_viscosity * viscosity_factor) +
-					(1 - viscosity_factor);
+				if (in_liquid_stable) {
+					if (dl > lplayer->movement_liquid_fluidity_smooth)
+						dl = lplayer->movement_liquid_fluidity_smooth;
+				} else if (dl > (0.5f * BS)) {
+					dl = (0.5f * BS);
```

Comment:
Why 0.5? Would you please be so nice to add a comment here?

Reply:
Also, please remove the unnecessary parentheses.